middle/src/lib.rs

```diff
@@ -1,36 +1,49 @@
-extern crate base;
+extern crate baseapi;
 
-use base::{Logger, LogEntry};
+use std::convert::AsRef;
+use baseapi::{LogEntryTrait, LoggerTrait};
 use std::mem;
 
-pub struct Middleware {
-    logger: Logger,
-    violations: Vec<LogEntry>,
+struct Logger;
+
+pub trait MiddlewareTrait<T:LoggerTrait<U> ,U:LogEntryTrait+AsRef<str> >{
+    type associatedType1 :LoggerTrait<U>;
+    fn new(logger: T)->Self;
+    fn log_violation(&mut self, s: &str); 
+    fn take_violations(&mut self) -> Vec<U>; 
+    fn take_logger(self) -> T;
 }
 
-impl Middleware {
-    fn new(logger: Logger) -> Middleware {
+
+pub struct Middleware<T:LoggerTrait<U>, U: LogEntryTrait+AsRef<str> > {
+    logger: T,
+    violations: Vec<U>,
+    }
+
+impl<T:LoggerTrait<U> ,U:LogEntryTrait+AsRef<str> > MiddlewareTrait<T,U> for Middleware<T,U> {
+    type associatedType1 = Logger;
```

Comment:
@jdm This line shows the following error:
impl<T:LoggerTrait<U> ,U:LogEntryTrait+AsRef<str> > MiddlewareTrait<T,U> for Middleware<T,U> {
   |                                                     ^^^^^^^^^^^^^^^^^^^^ the trait `baseapi::LoggerTrait<U>` is not implemented for `Logger`

I wanted to get rid of generic traits T and U. For this I created associatedType1 for testing, but it shows error that it needs implementation.

Reply:
Sorry, I was imprecise when I suggested using associated types to remove the repeated generic types everywhere. We don't get any value from defining a MiddlewareTrait in the middle/ crate that also defines the Middleware type. If we were to add a middle_api crate for it then it would make more sense, but I would not focus on trying to do any splitting of middle/ until the rest of the changes work correctly. Minimizing the amount of change at any given time is the best strategy when prototyping splitting crates.

Reply:
@jdm  Thanks alot for the taking out time and explaining me. So what do you ask me to do ahead? Which generic traits should I try get rid off?